### PR TITLE
KEP 07 Implementation

### PR DESCRIPTION
**Pull request by jlpedrosa**
**What this PR does / why we need it**:
This implements KEP 07 (Draft) by appending the AssertCommands to the TestStep structure.

```go
type TestAssertCommand struct {
	// The command and argument to run as a string.
	Command string `json:"command"`
	// If set, the `--namespace` flag will be appended to the command with the namespace to use.
	Namespaced bool `json:"namespaced"`
	// Ability to run a shell script from TestStep (without a script file)
	// namespaced and command should not be used with script.  namespaced is ignored and command is an error.
	// env expansion is depended upon the shell but ENV is passed to the runtime env.
	Script string `json:"script"`
	// If set, the output from the command is NOT logged.  Useful for sensitive logs or to reduce noise.
	SkipLogOutput bool `json:"skipLogOutput"`
}
```

There's no issue associated with KEP 07.

> Create TestAssertCommand struct, and add it as a list to TestAssert
> Create Helper functions in utils to run assertions'
> Update RunCommands to receive an existing context
> Update step Check method to also run the scripts/commands specified in the step assertion
> Update the set Run command to honor the timeout (as commands can be lengthy)

### CODE REVIEW

File: pkg/test/utils/kubernetes.go
Comment (kensipe) [inline pkg/test/utils/kubernetes.go:1076]: This does not below in `kubernetes.go` if it goes in `utils` at all...  `utils` predates you (and me) and would prefer better packaging... don't let that block you... but it shouldn't be in `kubernetes`.   This is more kuttl specific behavior

**Review by kensipe — CHANGES_REQUESTED** [review]

File: pkg/test/utils/kubernetes.go
Comment (jlpedrosa) [reply pkg/test/utils/kubernetes.go:1076]: I can move that function to the file you tell me. the thing is `kubernetes.go` is already coupled with "harness" namespace and all the utils are using the data model from API. If it would be a 1:1 mapping I would suggest yo move it to `api` package, but since I'm adding some "logic" (background false, timeout..) to make it compatbible with the existing method signatures, I am not clear what would be the "right" place. Maybe in Steps.go. 
 My natural automatic reaction would be make a pakage with all the run utils and it's own data structures, but that is clearly beyond the scope of this PR.
Let me know what is your choice.

Comment (kensipe) [reply pkg/test/utils/kubernetes.go:1076]: yeah... I need to dig in.... I was hoping not to add any more coupling there.  It's not too significant... there is a lot of refactoring I would like to do... I don't want to place that burden on you or this PR

Comment (kensipe) [inline pkg/test/utils/kubernetes.go:1076]: I had not thought of the solution this way... but makes sense.
what doesn't make sense is the method name `toAssertCommands` it isn't `to` assert it is `from` assert commands and to standard commands.    Naming is hard but this one is misleading.   perhaps `convertAssertCommand`

File: pkg/test/utils/kubernetes.go
Comment (kensipe) [inline pkg/test/utils/kubernetes.go:1086]: misspelling `constans`

File: pkg/test/step.go
Comment (kensipe) [inline pkg/test/step.go:361]: In order to understand this, it is necessary to dig deep.. it could be useful to add a comment that the errors is a combination of errors to execute (command doesn't exist) AND exit error codes.    Just a suggestion... otherwise looks good.

File: pkg/test/step.go
Comment (kensipe) [inline pkg/test/step.go:416]: this needed some attention.. is there a reason to change it on this PR?  could you open this separately please.

**Review by kensipe — CHANGES_REQUESTED** [review]
/lgtm. Nice @jlpedrosa   The only thing stopping this is tests... we need to create tests which are often serve as an example of how to do things.

consider adding to:  https://github.com/kudobuilder/kuttl/tree/main/pkg/test/step_integration_test_data

File: pkg/test/utils/kubernetes.go
Comment (jlpedrosa) [reply pkg/test/utils/kubernetes.go]: Renamed

File: pkg/test/utils/kubernetes.go
Comment (jlpedrosa) [reply pkg/test/utils/kubernetes.go]: Corrected

File: pkg/test/step.go
Comment (jlpedrosa) [reply pkg/test/step.go]: Comment added

File: pkg/test/step.go
Comment (jlpedrosa) [reply pkg/test/step.go:420]: Let me explain why this is I consider this necessary in this PR. As commands for assertions most likely will be far less than instant, the previous implementations won't respect the timeout at all. Let's assume the command takes 10 seconds to execute, and the timeout is 30 seconds, this means that the loop would be running for 330 seconds (10*30 + 1*30) as each loops executes it (10 secs) and just waits 1 second.

Comment (kensipe) [reply pkg/test/step.go:420]: yeah... I've been frustrated with the "meaning" of timeout and our inconsistent use of timeout for awhile.  I imagined it needed to be address.  I just thought we would have some discussion around it.. etc.    The problem with it in this PR is it is an important component lost in the noise of another important component.  It establishes an opinion which should be debated a bit.

Things to figure out:
1. what does timeout apply to... it could be the testsuite timeout, it could be a test, or a test step.  It could be each independent element in the teststep...
2. The timeout could be separate for the apply phase and assert phase or be inclusive
3. When a timeout change is made it is common to have it made for the assert phase (with knowledge that it takes extra time for the apply to be realized in the assert).  Does that affect assert only.  Is it all asserts collectively.

I also imagine that AssertCommands, might make sense to have a timeout themselves (which is currently missing).  There is the length of time allow for a successful invocation, there is the number of times you are willing to make the invocation (defined here.. as the number of times it can be called within an assert timeout.).

Timeouts can be very important for eventually consistent tests... I believe we have a number of timeouts conflated conceptually together and we need to clean up what a timeout means at different levels of usage.

I will look past it here.  Your position makes sense.   I'm really excited with contributors add value to through PRs as you have and look to provide guidance... but also want to make sure the burden isn't high.  Thanks for the contribution!

File: pkg/test/step_integration_test.go
Comment (kensipe) [inline pkg/test/step_integration_test.go:509]: our linter must not be picking up the integration tests... but these are not properly formatted

**Review by kensipe — CHANGES_REQUESTED** [review]
This is close... it isn't properly formatted for the integration tests... otherwise I'm good.  It would be great to get @alenkacz review as well

I will look to understand how to make sure integration builds are included in the linting

File: pkg/test/step_integration_test.go
Comment (kensipe) [reply pkg/test/step_integration_test.go:509]: @jlpedrosa making the mod temporarily to the make from this PR will help ensure alignments on tests.  https://github.com/kudobuilder/kuttl/pull/274

**Comment by kensipe** [comment]
@jlpedrosa is this still  WIP?  or ready for review?  Committers are more likely to review if WIP is complete.  thanks again!

File: pkg/test/step.go
Comment (alenkacz) [inline pkg/test/step.go:423]: I need to take a closer look at this, I'll circle back to this on Monday. I would expect we have a context set with that timeout and then we pass context around also including all the methods.

Maybe this is valid implementation as well and we can do the context one as follow-up though.

**Review by alenkacz — CHANGES_REQUESTED** [review]
Generally looks good. Please bare with me with regards to the timeout stuff :)

File: pkg/test/step.go
Comment (jlpedrosa) [reply pkg/test/step.go:361]: I agree, but it was inconsistent with all the other `ChecksXXXX`  that were public ie: `CheckResource` `CheckResourceAbsent`.  Let me know you decision, and I'll adapt the code.

File: pkg/test/step.go
Comment (jlpedrosa) [reply pkg/test/step.go:423]: I think that would be better, and that is the direction I was going (that is why I added ctx to many methods). But also i was trying not to do too much code changes to try to adhere to @kensipe guidelines. My reasoning was: if it's a good practice and it has minimal  or no impact, the do it. (That is why adding the context fell in).

File: pkg/test/step.go
Comment (jlpedrosa) [reply pkg/test/step.go:420]: I think I get your point,  but I would say the modification of the code, through maybe opinionated, has a very small scope and it's not taking a big step in any direction. In other words, what ever discussion and design choice made later does not get impacted by applying this PR or not, the diff of this code and the previous one is minimal. Of course that is my opinion from outside,  what ever you guys decide I'll get it implemented.

File: pkg/test/utils/kubernetes.go
Comment (alenkacz) [inline pkg/test/utils/kubernetes.go:1112]: just wondering what motivated this change? Was it that originally you planned to pass in context from the top but in the end did not end up doing it and considered it out of scope of this PR?

File: pkg/test/step.go
Comment (alenkacz) [inline pkg/test/step.go:369]: let's also not that we run commands now

File: pkg/test/step.go
Comment (alenkacz) [inline pkg/test/step.go:361]: command -> commands (it's an array)

**Review by alenkacz — APPROVED** [review]
Thanks, in the end I have only couple of nits. I want to take a closer look at the timeouts after this merges anyway. With timeout on step level and also on the command level and context being passed in but not containing any timeouts, it's a bit not straightforward but that's not your fault :)

Thanks for the great contribution 👏

File: pkg/test/step_integration_test.go
Comment (jlpedrosa) [reply pkg/test/step_integration_test.go:509]: @kensipe  if you could merge #274 I could rebase mine and run the test. If I cherry pick that commit here we would have conflicts no?

Comment (alenkacz) [reply pkg/test/step_integration_test.go:509]: merged :)

File: pkg/test/utils/kubernetes.go
Comment (jlpedrosa) [reply pkg/test/utils/kubernetes.go:1112]: Yes, I was preparing to do bigger changes on the timeouts (as you suggested) but the other conversations in the ticket seemed that it would be too big of a change, so I left what made sense on it's own and was a good golang practice (pass the context).
NExt week I could give another hand for the global timeouts if you would like.

File: pkg/test/step_integration_test.go
Comment (jlpedrosa) [reply pkg/test/step_integration_test.go:504]: rebased, and tests fmtd.

File: pkg/test/utils/kubernetes.go
Comment (alenkacz) [reply pkg/test/utils/kubernetes.go:1112]: Sounds great, thank you :)

File: pkg/test/step.go
Comment (jlpedrosa) [reply pkg/test/step.go]: Fixed.

Comment (jlpedrosa) [reply pkg/test/step.go]: fixed.

File: pkg/test/step.go
Comment (jlpedrosa) [reply pkg/test/step.go:369]: I am not sure what you ment here. I've updated the comment as what I read that the comment was not explaining that we also run the commands here.

File: pkg/test/step.go
Comment (kensipe) [reply pkg/test/step.go:423]: one of the timeouts... should be passed thru with context... completely agree with @alenkacz .   not holding this PR to that... but is the right way to manage 1 or 2 of the timeouts.  a topic which should tackle next if possible.

**Review by kensipe — APPROVED** [review]
/lgtm

Good stuff!  Thanks for making kuttl better!!

**Pull request merged**
